### ISSUE TITLE
머딥 - 13.5d (Scaling 심화, 구간분할 예측)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#44]
# 목차
>1. Scaling 간략한 소개
>2. fit_transform

## 소개
1. 한 쪽 방향으로만 outlier가 많으면 StandardScaler

2. 양 쪽 다 outlier가 많으면 RobustScaler

3. Normalizer (코사인 유사도 용)
: 유클라디언 거리를 모두 1로 만든다

**코사인 유사도**
1 = 완벽하게 동일한 데이터
-1 = 완벽하게 반대되는 데이터

## fit_transform
```sage
from sklearn.preprocessing import MinMaxScaler
from sklearn.datasets import load_breast_cancer 
from sklearn.model_selection import train_test_split

cancer = load_breast_cancer()
X_train, X_test, y_train, y_test = train_test_split(
    cancer['data'],
    cancer['target'],
    random_state=0
    )

from sklearn.svm import SVC
minmax = MinMaxScaler()
```

#### 각 특성마다 최소, 최대값 계산(아직 안 바뀌었음)
```sage
X_train_scaled = minmax.fit_transform(X_train)
```
#### train 데이터 기준으로 X_max와 X_min이 결정되어 있음
```sage
X_test_scaled = minmax.transform(X_test)
```

### score 함수 쓰면 pred가 필요없다. valid넣으면 알아서 계산해줌!!!!!
```sage
svm = SVC(C=100)
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

svm2 = SVC(C=100)
svm2.fit(X_train_scaled, y_train)
svm2.score(X_test_scaled, y_test) # 2% 성능향상
```

**Comment by vmtmxmf5** [comment]
https://jhryu1208.github.io/data/2021/01/11/ML_segmentation/

#### 구간분할
```sage
import numpy as np
from sklearn.preprocessing import OneHotEncoder
```
#### datapoint의 소속 구간
```sage
which = np.digitize(X, bins=np.linspace(-3, 3, 11))
encoder = OneHotEncoder(sparse=False) #희소행렬로 안만들기
encoder.fit(which)

X_binned = encoder.transform(which)
```

## 구간분할 왜 쓰는가?
1. 계층 분할
2. 구간 예측

구간정보를 부여했을 경우 연속형 자료로 쓰는 것보다 어떤 점이 우월한가? 
**선형 모델은 더 유연**하게, **트리 모델은 덜 유연(부드럽게)**하게 만들 수 있다.

(트리 모델 입장에서는 덜 유연해지는데, 그 이유는 구간 데이터가 원핫인코딩으로 1과 0 밖에 없기 때문이다)

### HOW??
feature 개수가 부족할 때, 구간분할을 사용해 **feature 개수**를 뻥튀기 + feature 의 **구간별 기울기** 예측

### So!
하나의 feature 를 이용해 y를 더 정확하게 예측할 수 있다



#### 구간별 평균
```sage
np.hstack([X_binned, X])
```
=> y평균이 나온다
x feature가 **하나**밖에 없기 때문에 구간별 기울기가 모두 같다.


### 구간에 들어간 실제 X값이 있다면?
=> 구간별 기울기를 따로 구할 수 있다

#### 구간별 실제 들어간 값
```sage
np.hstack([X_binned, X * X_binned])
```
X는 1개의 열, X_binned는 원핫인코딩 10개 열.
즉, 브로드캐스팅으로 계산하면 좌측 10열은 구간, 우측 10열은 구간에 해당하는 X값
